ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CreateFunctionCommand.java

```diff
@@ -0,0 +1,228 @@
+    final List<Schema> arguments = new ArrayList<>();
+    for (TableElement element : elements) {
+      arguments.add(LogicalSchemas.fromSqlTypeConverter().fromSqlType(element.getType()));
+    }
+    return arguments;
+  }
+
+  private static String[] elementsToParamNames(final List<TableElement> elements) {
+    final int size = elements.size();
+    final String[] paramNames = new String[size];
+    for (int i = 0; i < size; i++) {
+      paramNames[i] = elements.get(i).getName();
+    }
+    return paramNames;
+  }
+
+  private static Class[] elementsToParamTypes(final List<TableElement> elements) {
+    final int size = elements.size();
+    final Class[] paramTypes = new Class[size];
+    for (int i = 0; i < size; i++) {
+      final Schema schema = LogicalSchemas
+          .fromSqlTypeConverter()
+          .fromSqlType(elements.get(i).getType());
+      paramTypes[i] = SchemaUtil.getJavaType(schema);
+    }
+    return paramTypes;
+  }
+
+  /**
+   * A method for retrieving a custom Kudf class. This class gets instantiated
+   * everytime a query is created with the custom function.
+   */
+  static Class<? extends Kudf> getKudfClass(final String language) {
+
+    class InlineJavaUdf implements Kudf {
+      private final IScriptEvaluator se;
+
+      @SuppressWarnings("checkstyle:redundantmodifier")
+      public InlineJavaUdf(
+          final CreateFunction createFunction,
+          final KsqlConfig ksqlConfig
+      ) throws Exception {
+        try {
+          // create a blacklist
+          final File pluginDir = new File(ksqlConfig.getString(KsqlConfig.KSQL_EXT_DIR));
+          final Blacklist blacklist = new Blacklist(new File(pluginDir, "resource-blacklist.txt"));
+
+          // create a class loader
+          final UdfClassLoader classLoader = UdfClassLoader.newClassLoader(
+              Optional.empty(),
+              Thread.currentThread().getContextClassLoader(),
+              blacklist);
+
+          // create a script executor
+          final List<TableElement> elements = createFunction.getElements();
+          se = CompilerFactoryFactory.getDefaultCompilerFactory().newScriptEvaluator();
+          se.setParentClassLoader(classLoader);
+          se.setReturnType(SchemaUtil.getJavaType(createFunction.getReturnType()));
+          se.setParameters(
+              elementsToParamNames(elements),
+              elementsToParamTypes(elements)
+          );
+          se.cook(createFunction.getScript());
+
+        } catch (CompileException ce) {
+          final String errorMessage = String.format("Failed to compile UDF");
+          ce.printStackTrace();
+          throw new KsqlException(errorMessage);
+        }
+      }
+
+      @SuppressWarnings("unchecked")
+      @Override
+      public Object evaluate(final Object... args) {
+        Object value;
+        try {
+          ExtensionSecurityManager.INSTANCE.pushInUdf();
+          value = se.evaluate(args);
```

Comment:
actually, this may not be the proper way to prevent a `System.exit`. Any ideas here?

Reply:
Looking at the code, this seems correct, though if we have to ask these questions then IMO we're structuring the code incorrectly. The security logic should definitely be a shared code path between builtin and inline udfs.